src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -29,21 +34,68 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
-            _factory.GetTempData(context.HttpContext).Save();
         }
 
         /// <inheritdoc />
         public void OnResultExecuting(ResultExecutingContext context)
         {
+            context.HttpContext.Response.OnStarting((state) =>
+            {
+                var saveTempDataContext = (SaveTempDataContext)state;
+
+                // If temp data was already saved, skip trying to save again as the calls here would potentially fail
+                // because the session feature might not be available at this point.
+                // Example: An action returns NoContentResult and since NoContentResult does not write anything to
+                // the body of the response, this delegate would get executed way late in the pipeline at which point
+                // the session feature would have been removed.
+                object obj;
+                if (saveTempDataContext.HttpContext.Items.TryGetValue(TempDataSavedKey, out obj))
+                {
+                    return TaskCache.CompletedTask;
+                }
+
+                SaveTempData(
+                    saveTempDataContext.ActionResult,
+                    saveTempDataContext.TempDataDictionaryFactory,
+                    saveTempDataContext.HttpContext);
+
+                return TaskCache.CompletedTask;
+            },
+            state: new SaveTempDataContext()
```

Comment:
You can avoid allocating an object here if you set the `ResultExecutingContext` as a property on the `Filter` and then passing `this`.